packages/blocks/src/database-block/database-block.ts

```diff
@@ -276,18 +360,154 @@ export class DatabaseBlockComponent
   @query(DATABASE_ADD_COLUMN_TYPE_POPUP)
   addColumnTypePopup!: DatabaseAddColumnTypePopup;
 
-  private _vEditor: VEditor | null = null;
-
   @query('.affine-database-block-title')
   private _container!: HTMLDivElement;
 
+  @query('.affine-database-search-input')
+  private _searchInput!: HTMLInputElement;
+
+  @state()
+  private _searchState: SearchState = SearchState.SearchIcon;
+
+  @state()
+  private _filteredRowIds: string[] = [];
+
+  @state()
+  private _hoverState = false;
+
+  private _vEditor: VEditor | null = null;
+  private _disposables: DisposableGroup = new DisposableGroup();
+
   get columns(): ColumnSchema[] {
     return this.model.columns.map(id =>
       this.model.page.getColumnSchema(id)
     ) as ColumnSchema[];
   }
 
+  connectedCallback() {
+    super.connectedCallback();
+    const disposables = this._disposables;
+
+    disposables.addFromEvent(this, 'mouseover', this._onMouseOver);
+    disposables.addFromEvent(this, 'mouseleave', this._onMouseLeave);
+    disposables.addFromEvent(this, 'click', this._onClick);
+  }
+
+  firstUpdated() {
+    this._initTitleVEditor();
+
+    this.model.propsUpdated.on(() => this.requestUpdate());
+    this.model.childrenUpdated.on(() => this.requestUpdate());
+  }
+
+  disconnectedCallback() {
+    super.disconnectedCallback();
+    this._disposables.dispose();
+  }
+
+  private _getRowKeyValues() {
+    // database paragraphs data (first column)
+    const rowKeyValues = this.model.children.reduce((acc, child) => {
+      return {
+        ...acc,
+        [child.id]: [child.text?.toString() ?? ''],
+      };
+    }, {} as RowData);
+
+    // columns data
+    const databaseData = this.model.page.columns.toJSON() as PageDatabaseData;
+    const databaseRowIds = Array.from(this.model.childMap.keys());
+    databaseRowIds.forEach(key => {
+      const columns = databaseData[key];
+      // database column may has no value
+      if (!columns) return;
+
+      const columnData = Object.keys(columns).map(
+        key => columns[key].value + ''
+      );
+      rowKeyValues[key].push(...columnData);
+    });
+
+    return rowKeyValues;
+  }
+
+  private _onSearch = (event: InputEvent) => {
+    const inputValue = (event.target as HTMLInputElement).value.trim();
+    this._searchState = SearchState.Searching;
+    if (inputValue === '') {
+      this._searchState = SearchState.SearchInput;
+    }
+
+    const rowKeyValues = this._getRowKeyValues();
```

Comment:
`_getRowKeyValues` is a heavy compted function, consider it to be a memo?

Reply:
Yes, I have encountered a problem before, after adding a new row, I can't get the content of the new row header column(first paragraph column).